src/metabase/db/connection_pool_setup.clj

```diff
@@ -1,17 +1,45 @@
 (ns metabase.db.connection-pool-setup
   "Code for creating the connection pool for the application DB and setting it as the default Toucan connection."
-  (:require [metabase.config :as config]
+  (:require [java-time :as t]
+            [metabase.config :as config]
             [metabase.connection-pool :as connection-pool]
             [schema.core :as s])
-  (:import com.mchange.v2.c3p0.PoolBackedDataSource))
+  (:import [com.mchange.v2.c3p0 ConnectionCustomizer PoolBackedDataSource]))
+
+(def ^:private latest-checkin (atom nil))
+
+(def ^:private recent-window-seconds 15)
+
+(defn recent-activity?
+  "Returns true if there has been recent activity. Define recent activity as an application db connection checked in
+  within 15 seconds. Check-in means a query succeeded and the db connection is no longer needed."
+  []
+  (when-let [activity @latest-checkin]
+    (t/after? activity (t/minus (t/offset-date-time) (t/seconds recent-window-seconds)))))
+
+(defrecord CheckinTracker []
+  ConnectionCustomizer
+  (onAcquire [_ _connection _identity-token])
+  (onCheckIn [_ _connection _identity-token]
+    (reset! latest-checkin (t/offset-date-time)))
+  (onCheckOut [_ _connection _identity-token]
+    (reset! latest-checkin (t/offset-date-time)))
```

Comment:
Was this left in unintentionally? Or maybe the docstring for `recent-activity?` and the atom name need tweaking?

Reply:
Which strikes you as best? I'm leaning towards making everything keyed off of check in but open to other ideas.